File: ping-pong-1/solution.go
package main

import "strings"

func PingPong(sounds string) string {
	soundList := strings.Split(sounds, "-")
	
	pingScore := 0
	pongScore := 0
	
	var currentServer string
	var lastValidHitter string
	var finalBadHitter string
	
	for _, sound := range soundList {
		if sound == "ping" || sound == "pong" {
			if currentServer == "" {
				currentServer = sound
			}
			lastValidHitter = sound
		} else {
			// bad noise marks end of rally
			if currentServer != "" {
				if currentServer != lastValidHitter {
					// server wins (opponent made the bad shot)
					if currentServer == "ping" {
						pingScore++
					} else {
						pongScore++
					}
				}
			}
			// track final bad hitter for tie-breaking
			finalBadHitter = lastValidHitter
			// reset for next rally
			currentServer = ""
			lastValidHitter = ""
		}
	}
	
	// determine winner
	if pingScore > pongScore {
		return "ping"
	} else if pongScore > pingScore {
		return "pong"
	} else {
		// tie: winner is who didn't hit the final bad shot
		if finalBadHitter == "ping" {
			return "pong"
		}
		return "ping"
	}
}

File: ping-pong-1/solution.js
function pingPong(sounds) {
  const hits = sounds.split('-');
  let server = null;
  let lastGoodHit = null;
  let inBadNoises = false;
  let finalBadShotBy = null;
  const scores = { ping: 0, pong: 0 };

  for (const hit of hits) {
    if (hit === 'ping' || hit === 'pong') {
      if (inBadNoises || server === null) {
        server = hit;
        inBadNoises = false;
      }
      lastGoodHit = hit;
    } else {
      if (!inBadNoises) {
        const badShotBy = lastGoodHit;
        const winner = badShotBy === 'ping' ? 'pong' : 'ping';
        if (winner === server) scores[winner]++;
        finalBadShotBy = badShotBy;
      }
      inBadNoises = true;
    }
  }

  if (scores.ping !== scores.pong) return scores.ping > scores.pong ? 'ping' : 'pong';
  return finalBadShotBy === 'ping' ? 'pong' : 'ping';
}